### PR TITLE
Add MatrixZero and use it in the LSTM

### PR DESCRIPTION
**Pull request by sffc**
Running benchmarks on this PR on my MacBook Air M2 shows no performance regression and perhaps a small performance boost, which is both great and the opposite of intuition. Would appreciate someone else running the benchmark before this change and after it:

```
$ cd experimental/segmenter
$ cargo bench --all-features lstm
```

### CODE REVIEW

**Comment by zbraniecki** [comment]
This PR generates a reliable 7-8% improvement on my MBP M1 Pro:

```
Line Break/UTF8/Th/lstm time:   [257.59 µs 257.87 µs 258.19 µs]
                        change: [-9.0029% -8.5994% -8.2038%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) low mild
  2 (2.00%) high severe

Line Break/UTF16/Th/lstm
                        time:   [255.78 µs 256.20 µs 256.67 µs]
                        change: [-7.2360% -7.0447% -6.8566%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```

**Review by robertbastian — DISMISSED** [review]
I like this.

Thought: I think these types are usually called tensors in ML (with matrix being a 2-tensor). Rename?

**Comment by sffc** [comment]
I verified that the 6-8% performance boost also occurs on x86_64.

**Comment by aethanyc** [comment]
I see performance improvement 6-7% on my MacBook Pro (Apple M1 Max).
```
Line Break/UTF8/Th/lstm time:   [252.84 µs 253.33 µs 253.91 µs]
                        change: [-8.0179% -7.7523% -7.4965%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

Line Break/UTF16/Th/lstm
                        time:   [253.58 µs 254.39 µs 255.24 µs]
                        change: [-6.5235% -6.1828% -5.8240%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
```


Also, a performance improvement 3-4% on my Linux desktop (Intel i5-9600K).

```
Line Break/UTF8/Th/lstm time:   [471.29 µs 473.33 µs 475.72 µs]                                    
                        change: [-3.5093% -3.0502% -2.5363%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

Line Break/UTF16/Th/lstm                                                                            
                        time:   [470.34 µs 471.96 µs 473.97 µs]
                        change: [-4.4536% -4.0915% -3.7435%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
```

File: experimental/segmenter/src/lstm_bies.rs [hunk 1]
```diff
@@ -52,25 +52,25 @@ impl<'l> Lstm<'l> {
         // The ICU4X style guide discourages this. We do it here because:
```
[the remark this answers is not in the document]
Comment (sffc) [reply experimental/segmenter/src/lstm_bies.rs]: Yes, I proved this comment wrong :)

**Comment by sffc** [comment]
> Thought: I think these types are usually called tensors in ML (with matrix being a 2-tensor). Rename?

_Maybe_, though tensors have their own special definition, too. For the time being I'll stick with the language I learned in my ML and linear algebra classes.

**Comment by robertbastian** [comment]
Wikipedia is pretty clear about matrices being 2d, and most matrix operations don't generalise to higher dimensions.

I think the `submatrix` method name is especially confusing, because depending on the generic, it returns a 2d matrix, a 1d vector, or even a 0d scalar.

**Comment by sffc** [comment]
Matlab and Octave call them "multidimensional arrays"

https://docs.octave.org/v4.2.2/Advanced-Indexing.html

https://www.mathworks.com/help/matlab/math/multidimensional-arrays.html

I think this is likely why ndarray calls them "arrays".

Here's a stack overflow post trying to explain the differences

https://math.stackexchange.com/questions/1134809/are-there-any-differences-between-tensors-and-multidimensional-arrays

But the problem is that "arrays" are an overloaded term because they mean something very specific in Rust.

**Comment by sffc** [comment]
I'm slightly inclined to merge this with the current naming because

1. We can change these names later
2. There's no clear great alternative

**Review by aethanyc — APPROVED** [review]

**Pull request merged**